storage/remote/codec.go

```diff
@@ -232,19 +234,24 @@ func StreamChunkedReadResponses(
 				continue
 			}
 
-			b, err := proto.Marshal(&prompb.ChunkedReadResponse{
+			resp := &prompb.ChunkedReadResponse{
 				ChunkedSeries: []*prompb.ChunkedSeries{
 					{Labels: lbls, Chunks: chks},
 				},
 				QueryIndex: queryIndex,
-			})
+			}
+
+			b, err := resp.PooledMarshal(marshalPool)
 			if err != nil {
 				return ss.Warnings(), fmt.Errorf("marshal ChunkedReadResponse: %w", err)
 			}
 
 			if _, err := stream.Write(b); err != nil {
 				return ss.Warnings(), fmt.Errorf("write to stream: %w", err)
 			}
+
+			// We immediately flush the Write() so it is safe to return to the pool.
```

Comment:
Did you double check it?

Reply:
Yes, here it is:
https://github.com/prometheus/prometheus/blob/main/storage/remote/read_handler.go#L222
https://github.com/prometheus/prometheus/blob/8553a98267a56acfafc56ba69ab8946b71655304/storage/remote/chunked.go#L89

A flusher is always provided there: https://github.com/prometheus/prometheus/blob/main/storage/remote/read_handler.go#L187.